data-prepper-plugins/s3-source/src/main/java/org/opensearch/dataprepper/plugins/source/S3SelectObjectWorker.java

```diff
@@ -0,0 +1,237 @@
+	static final String S3_SELECT_OBJECT_FAILED_METRIC_NAME = "s3SelectObjectsFailed";
+	static final String S3_SELECT_EVENTS = "s3SelectObjectsEvents";
+	static final String S3_BUCKET_NAME = "bucket";
+	static final String S3_OBJECT_KEY = "key";
+	static final String S3_BUCKET_REFERENCE_NAME = "s3";
+
+	private final DistributionSummary s3SelectSizeProcessedSummary;
+	private final DistributionSummary s3SelectEventsSummary;
+	private final Counter s3SelectFailedCounter;
+	private final S3AsyncClient s3AsyncClient;
+	private final S3SourceConfig s3SourceConfig;
+	static final Map<String,Integer> retryLogicKeyWiseCounterMap = new HashMap<>();
+
+	public S3SelectObjectWorker(final S3AsyncClient s3AsyncClient, final S3SourceConfig s3SourceConfig,
+			final PluginMetrics pluginMetrics) {
+		this.s3SourceConfig = s3SourceConfig;
+		this.s3AsyncClient = s3AsyncClient;
+		s3SelectSizeProcessedSummary = pluginMetrics.summary(S3_SELECT_OBJECTS_SIZE_PROCESSED);
+		s3SelectEventsSummary = pluginMetrics.summary(S3_SELECT_EVENTS);
+		s3SelectFailedCounter = pluginMetrics.counter(S3_SELECT_OBJECT_FAILED_METRIC_NAME);
+	}
+	
+	public void selectObjectFromS3(final S3ObjectReference s3ObjectReference,
+			final BufferAccumulator<Record<Event>> bufferAccumulator) throws Exception {
+		final long totalBytesRead;
+		final InputStream inputStreamList;
+		final InputSerialization inputSerialization = getInputSerializationFormat(
+				s3SourceConfig.getS3SelectOptions().getDataSerializationFormat());
+		final SelectObjectContentRequest s3Request = getS3SelectObjRequest(s3ObjectReference, inputSerialization);
+		final S3SelectResponseHandler responseHandler = new S3SelectResponseHandler();
+		final S3SelectResponseHandler responseHand = callS3SelectObject(s3Request, responseHandler, s3ObjectReference,
+				bufferAccumulator);
+		List<SelectObjectContentEventStream> receivedEvents = responseHand.getReceivedEvents();
+		if (!receivedEvents.isEmpty()) {
+				inputStreamList = getInputStreamFromResponseHeader(responseHand);
+				totalBytesRead = parseCompleteStreamFromResponseHeader(s3ObjectReference, bufferAccumulator,
+						inputStreamList);
+				s3SelectSizeProcessedSummary.record(totalBytesRead);
+				s3SelectEventsSummary.record(bufferAccumulator.getTotalWritten());
+		}
+	}
+
+	private InputSerialization getInputSerializationFormat(final String format) {
+		InputSerialization inputSerialization = null;
+
+		if (format == null || format.isBlank()) {
+			throw new S3SelectException(
+					"data_serialization_format is mandatory configuration if s3 select query defined in pipeline configuration");
+		}
+		S3SelectSerializationFormatOption serializationFormat = S3SelectSerializationFormatOption
+				.valueOf(format.toUpperCase());
+		switch (serializationFormat) {
+		case CSV:
+			inputSerialization = InputSerialization.builder()
+					.csv(csv -> csv.fileHeaderInfo(FileHeaderInfo.USE).build())
+					.compressionType(CompressionType.NONE).build();
```

Comment:
We should allow the pipeline authors to specify the compression type.

Some buckets will have uncompressed CSV files and others will have compressed CSV files. The same goes for JSON and Parquet.

Please add this as a configuration in `S3SelectOptions`.

Reply:
Thanks, David for your review. we are already working on it. this initial code review doesn't include compression functionality. it will be available in the next PR.